wait_for_connection: Wait for system to become reachable

dagwieers opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New Module Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
wait_for_connection

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
v2.3

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->
This action plugin allows to check when a system is back online and usable by Ansible.

As an example, when doing a SysPrep and running Enable-WinRM.ps1, it takes between 10 to 20 seconds between the WinRM TCP port to open, and it actually being able to serve Ansible requests. This time is variable and depends on the boot process.

Current implementation should work for all connection types using the **ping** module (or **win_ping** on winrm/powershell)

This fixes #19998

dagwieers committed:
WIP: wait_for_connection: Wait for system to be reachable

This action plugin allows to check when a system is back online and
usable by Ansible.

As an example, when doing a SysPrep and running Enable-WinRM.ps1, it
takes between 10 to 20 seconds between the WinRM TCP port to open, and
it actually being able to server Ansible requests. This time is variable
and depends on the boot process.

Current implementation is specific for Windows (WinRM) only, this will
be fixed shortly.

This fixes #19998

dagwieers commented:
I don't particularly like the (long) name, but it's the best I could come up with :-/

dagwieers commented:
PS I would like to add the specific connection-related stuff (e.g. connect-test, ping-test) as part of the connection plugin itself. This way new connection methods can implement their own methodology.

dagwieers commented:
I am also contemplating a 'wait to become unreachable' and a 'wait for becoming reachable again' modus.
Again, not sure what terminology would be best. present/absent feels a bit odd here.

bcoca commented:
bikeshedding: 'connect'

bcoca commented:
The more i think of this, the more it seems like an expansion of ping

dagwieers commented:
bcoca: I agree, but two concerns:
- What would the interface look like ? Ping is so nice in its simplicity.
- People will probably not find this hidden in ping, most will do something sub-optimal with `wait_for` instead.

dagwieers committed:
Support other transport types

dagwieers commented:
ready_for_review

mattclay commented on test/integration/targets/wait_for_connection/aliases:
```diff
@@ -0,0 +1,2 @@
+posix/ci/group2
+windows/ci/group2
```
Missing newline at end of file.

dagwieers replied:
I don't really see why that is important. That is the default in VIM.
Ironically PEP8 is requiring me to remove the trailing newline, and you are asking me to add one :-)

mattclay replied:
PEP 8 requires a newline at the end of python files. Perhaps you're confusing it with a blank line?

As for the rational, it's about making clean diffs for future PRs. If I need to an another line at the end of the file, I need to modify the last line to add a newline. If it already has a newline, that's not necessary.

dagwieers replied:
Well, a newline at the end of my file is a blank line in my editor.

mattclay commented on test/integration/targets/wait_for_connection/tasks/main.yml:
```diff
@@ -0,0 +1,12 @@
+    connect_timeout: 5
+    sleep: 1
+    timeout: 10
+  connection: local
```
Missing newline at end of file.

dagwieers replied:
I don't really see why that is important. That is the default in VIM.
Ironically PEP8 is requiring me to remove the trailing newline, and you are asking me to add one :-)

mattclay replied:
Same as above.

dagwieers replied:
Done.

dagwieers commented:
Something wrong with Shippable ?

mattclay commented:
CI was failing due to changes to unit tests which passed CI on the PR but failed after merge. That change has been reverted and affected PRs are being run through CI again.

jhawkesworth commented on lib/ansible/modules/utilities/logic/wait_for_connection.py:
```diff
@@ -0,0 +1,64 @@
+DOCUMENTATION = '''
+---
+module: wait_for_connection
+short_description: Waitis until remote system is reachable/usable
```
Tiny typo here 'Waitis' -> 'Waits'

dagwieers replied:
Thanks ! Fixed.

dagwieers committed:
Various improvements

- Fix reported typo
- Add transport_test support in accelerate plugin
- Ensure port is an integer

dagwieers commented:
CI failed with TIMEOUT :-/

mattclay commented:
CI failure due to:

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: AttributeError: 'Connection' object has no attribute '_winrm_host'
```

dagwieers committed:
Improve examples

dagwieers commented:
Close and reopen because Shippable was UNSTABLE ?!?

jhawkesworth commented:
I think this is going to be very useful!

shipit

mattclay commented:
@dagwieers The unable status means that tests passed, but only after an automatic retry. This usually indicates a network error, such as failure to download a package required for a test.

dagwieers commented:
ready_for_review

jhawkesworth commented on lib/ansible/modules/utilities/logic/wait_for_connection.py:
```diff
@@ -0,0 +1,93 @@
+      customization:
+        hostname: '{{ vm_shortname }}'
+        runonce:
+        - powershell.exe -ExecutionPolicy Unrestricted -File C:\Windows\Temp\Enable-WinRM.ps1 -CertValidityDays 3650 -ForceNewSSLCert
```
I know you aren't fond of the current name, but for clarity I'd consider referring to the ConfigureRemotingForAnsible.ps1 name here

dagwieers replied:
Oh right, the example was written during the name-discussions. Will rename !

dagwieers committed:
Small fixes

- Use correct ConfigureRemotingForAnsible.ps1 script name
- Only use win_ping when remote shell is known to be Powershell
- Add integration tests to CI framework

dagwieers commented:
ready_for_review
bot_status

mattclay commented:
CI failure on wait_for_connection for FreeBSD for the chroot test. I'm not sure why this happens, since ansible ping (and other modules) work just fine for the same connection:

```
2017-02-17 00:50:06 fatal: [chroot-pipelining]: FAILED! => {
2017-02-17 00:50:06     "changed": false, 
2017-02-17 00:50:06     "elapsed": 600, 
2017-02-17 00:50:06     "failed": true, 
2017-02-17 00:50:06     "msg": "timed out waiting for ping module test success: 'ping'"
2017-02-17 00:50:06 }
```

dagwieers commented:
No CI today ?

dagwieers commented:
It's the second time I see that an amended commit does not start CI. In fact, there is no CI entry in the summary below. Closing and re-opening the PR does work.

mattclay commented:
Shippable is deploying a fix for the lost webooks issue this weekend. That should solve the missing CI problem.

dagwieers commented:
@mattclay Ah, I didn't know this symptom was related to that. Sorry for the noise. I'll refrain from sending you these bits until the dust has settled :-) Thanks again for your help and feedback !

dagwieers commented:
@mattclay It seems that the issue with FreeBSD was fixed by adding `task_vars` to the `_execute_module()` call. This was really unnecessary because it's the ping-module (with no module_args anyway). But somehow FreeBSD is affected by this.

dagwieers commented:
ready_for_review

dagwieers commented:
ready_for_review
shipit

mattclay commented:
Closing and re-opening to trigger CI.

dagwieers commented:
@gundalow @nitzmahone @mattclay @jhawkesworth ready_for_review

jhawkesworth commented:
shipit

nitzmahone commented on lib/ansible/plugins/connection/winrm.py:
```diff
@@ -86,6 +87,15 @@ def __init__(self,  *args, **kwargs):
 
         super(Connection, self).__init__(*args, **kwargs)
 
+        self.host = self._play_context.remote_addr
```
Do these values even need to exist in most of the plugins? Only reason I can see that they'd need to is if `transport_test` were defined on the base connection (and override-able by plugins that don't use a TCP port). Otherwise, `transport_test` can just use the existing `self._winrm_(host/port)`. If they do stay, I'd prefer that these values be init'd in `set_host_overrides` off the existing values instead of duplicating the logic here (at least for this plugin- IIRC nobody else uses `set_host_overrides` ATM). I've verified that it works with that change (since you're using the play's existing connection, it's already been run by the time it gets there).

dagwieers replied:
To be honest I preferred some consistency in the various connection plugins. But I really don't mind.
The difference now looks more better self-contained, so I am pleased with that as well.

nitzmahone commented on lib/ansible/plugins/action/wait_for_connection.py:
```diff
@@ -0,0 +1,119 @@
+            result['msg'] = str(e)
+
+        elapsed = datetime.now() - start
+        result['elapsed'] = elapsed.seconds
```
`elapsed_sec`/`'elapsed_seconds`? I personally dislike time values/args without units...

dagwieers replied:
I used the same output as wait_for. That was the intent.

dagwieers replied:
The return values are now documented, and the units are specified there as well.

nitzmahone reviewed (changes requested):
A couple of minor things... Unless this stock version of `transport_test` gets moved to the connection base (which isn't a bad idea IMO- no sense in duplicating what most connection plugins need), I think a number of the host/port members you added can probably go away.

dagwieers commented:
@gundalow @nitzmahone @mattclay @jhawkesworth Can we get this merged before v2.3 freeze ?

jhawkesworth commented:
Still looks good to me.  I guess @nitzmahone will want to re-check.

In case its been invalidated by subsequent commits

shipit

dagwieers commented:
Please merge before 12AM  :-)
